x11rb-protocol/src/parse_display/connect_instruction.rs

```diff
@@ -0,0 +1,218 @@
+        ConnectInstruction {
+            parsed_display,
+            stage: Stage::default(),
+        }
+    }
+
+    /// We should try to connect to the hostname and port over TCP.
+    fn try_hostname_over_tcp(&self) -> bool {
+        let ParsedDisplay { host, protocol, .. } = self.parsed_display;
+
+        !matches!(
+            (host.as_str(), protocol.as_deref()),
+            ("", _) | ("unix", _) | (_, Some("unix"))
+        )
+    }
+
+    /// We should try to connect to a socket.
+    fn try_unix_socket(&self) -> bool {
+        let ParsedDisplay { protocol, .. } = self.parsed_display;
+
+        cfg!(unix) && matches!(protocol.as_deref(), None | Some("unix"))
+    }
+
+    /// We should try to connect over the port.
+    fn try_port(&self) -> bool {
+        let ParsedDisplay { protocol, host, .. } = self.parsed_display;
+
+        protocol.is_none() && host.is_empty()
+    }
+
+    /// Get the hostname/port pair to connect to.
+    fn hostname_port_pair(&self) -> (&'a str, u16) {
+        let ParsedDisplay { host, display, .. } = self.parsed_display;
+
+        (host.as_str(), display + TCP_PORT_BASE)
+    }
+
+    /// Get the socket to connect to.
+    fn socket_path(&self) -> PathBuf {
+        let ParsedDisplay { display, .. } = self.parsed_display;
+
+        let path = format!("/tmp/.X11-unix/X{}", display);
+
+        // convert to a PathBuf
+        path.into()
+    }
+}
+
+impl<'a> Iterator for ConnectInstruction<'a> {
+    type Item = ConnectAddress<'a>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            // match on the current stage and try it
+            match self.stage {
+                Stage::TryHostname => {
+                    self.stage = Stage::TrySocket;
+                    if self.try_hostname_over_tcp() {
+                        let (hostname, port) = self.hostname_port_pair();
+
+                        // once we've tried the hostname, don't try anything else
+                        self.stage = Stage::OutOfOptions;
+
+                        return Some(ConnectAddress::Hostname(hostname, port));
+                    }
+                }
+                Stage::TrySocket => {
+                    self.stage = Stage::TryPort;
+                    if self.try_unix_socket() {
+                        let path = self.socket_path();
+                        return Some(ConnectAddress::Socket(path));
+                    }
+                }
+                Stage::TryPort => {
+                    self.stage = Stage::OutOfOptions;
+                    if self.try_port() {
+                        let port = self.parsed_display.display + TCP_PORT_BASE;
+                        return Some(ConnectAddress::Hostname("localhost", port));
+                    }
+                }
+                Stage::OutOfOptions => {
+                    return None;
+                }
+            }
+        }
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
```

Comment:
Heh, do we need a size hint? The default implementation of `(0, None)` might not be the tightest bound, but it should still be good enough for 99.99% of uses of this API.

Reply:
Or maybe it could have a simpler implementation that just returns something constant like `(0, Some(3))`.